File: internal/stats/account.go
package stats

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/Yata-Dash/Yata-Dash/internal/models"
)

// Account-warning fields — the two deadlines a private-tracker account can
// quietly miss.
//
// Both are the same shape: a stored timestamp, a clock, and a number of days.
// Neither announces itself. An expired API key means stats stop arriving and
// the tracker just looks broken; a pruned account means everything the ladder
// was built toward is gone. Both are trivially avoidable before the date and
// irreversible after it.
//
// These are derived on every read rather than stored, because their inputs are
// a timestamp and "now" — storing them would mean a value that is wrong by one
// day for every day since it was written.
const (
	// FieldLastLogin is the "last login" timestamp. Source-agnostic by design:
	// whatever writes it — a tracker's API, or the user's own "I've logged in"
	// — drives everything below without further changes.
	//
	// It follows the ORDINARY merge priority, with no special case: the
	// tracker's API is the source of truth about the tracker's own account,
	// and the user's record only fills in where the API says nothing. That is
	// almost everywhere — of 22 trackers probed on 2026-09-02, one reported a
	// login time — so the manual value is the live one in practice while the
	// two trackers that do report defer to what the tracker itself says.
	FieldLastLogin = "last_login"
	// FieldDaysSinceLogin is how long ago that was, in whole days.
	FieldDaysSinceLogin = "days_since_login"
	// FieldLoginDaysRemaining is how long is LEFT before the tracker acts on
	// the account, from its declared max_login_gap_days. Negative = overdue.
	FieldLoginDaysRemaining = "login_days_remaining"
	// FieldAPIKeyExpiresAt / FieldAPIKeyExpiryDays are the same pair for a key
	// with an expiry date.
	FieldAPIKeyExpiresAt  = "api_key_expires_at"
	FieldAPIKeyExpiryDays = "api_key_expiry_days"
)

// AccountPolicy carries the def-declared account rules for one tracker — the
// part of this calculation the stats engine cannot know on its own, since a
// tracker's inactivity policy comes from its def rather than from its stats.
type AccountPolicy struct {
	// MaxLoginGapDays is the tracker's inactivity deadline; 0 = not known.
	MaxLoginGapDays int
}

// deriveAccountFields adds the days-remaining fields to a merged view.
//
// Every field here is OMITTED when its input is missing, never emitted as a
// helpful-looking zero. That distinction is the whole feature: alert
// conditions treat an absent field as "no answer" and never match it, so
// silence on the twenty trackers that report nothing is automatic — whereas a
// days_since_login of 0 would read as "logged in today" on all of them, which
// is both false and exactly backwards.
func (e *Engine) deriveAccountFields(trackerID string, out models.MergedStats, now time.Time) {
	if src, ok := out[FieldLastLogin]; ok {
		if t, ok := statTime(src.Value); ok {
			since := wholeDays(now.Sub(t))
			out[FieldDaysSinceLogin] = derivedFrom(src, since)
			// The deadline needs the tracker's policy, which only a def can
			// supply. Without one there is nothing to count down to — and
			// saying so by omission is right: Yata not knowing a policy is not
			// evidence that the tracker has none.
			gap := 0
			if e.AccountPolicy != nil {
				gap = e.AccountPolicy(trackerID).MaxLoginGapDays
			}
			if gap > 0 {
				deadline := t.AddDate(0, 0, gap)
				out[FieldLoginDaysRemaining] = derivedFrom(src, wholeDays(deadline.Sub(now)))
			}
		}
	}
	if src, ok := out[FieldAPIKeyExpiresAt]; ok {
		if t, ok := statTime(src.Value); ok {
			out[FieldAPIKeyExpiryDays] = derivedFrom(src, wholeDays(t.Sub(now)))
		}
	}
}

// derivedFrom builds a derived field carrying the provenance of the timestamp
// it was computed from, so the UI's source dot points at where the underlying
// date actually came from rather than inventing a source of its own.
func derivedFrom(src models.StatField, days int) models.StatField {
	return models.StatField{Value: days, Source: src.Source, UpdatedAt: src.UpdatedAt}
}

// wholeDays truncates a duration toward zero in days: 25 hours ago is "1 day",
// 3 hours ago is "0 days" (today), and a deadline 6 hours away is "0 days"
// rather than a reassuring "1".
func wholeDays(d time.Duration) int { return int(d.Hours() / 24) }

// statTimeLayouts are the timestamp formats these fields actually arrive in.
//
// The bare "2006-01-02 15:04:05" form (Anthelion's LastAccess) carries NO
// timezone and is read as UTC, which can be up to 14 hours out. That is
// acceptable here and would not be everywhere: these thresholds are measured
// in days to weeks, so half a day of slop cannot flip a warning that fires a
// week ahead of its deadline. Contrast the freeleech countdown, where a
// mis-resolved "EST" put the end time five hours out on a display counting
// hours and minutes, and the abbreviation had to be resolved properly (see
// namedZoneOffsets in internal/fetch). Same trap, different tolerance.
var statTimeLayouts = []string{
	time.RFC3339,
	"2006-01-02 15:04:05",
	"2006-01-02T15:04:05",
	"2006-01-02",
}

// statTime reads a stored timestamp. Accepts the RFC3339 most APIs use, a
// space-separated variant, a bare date, and epoch seconds — a stat layer holds
// whatever JSON the tracker sent, so the number cases are not hypothetical.
func statTime(v any) (time.Time, bool) {
	switch t := v.(type) {
	case float64:
		return unixOrZero(int64(t))
	case int64:
		return unixOrZero(t)
	case int:
		return unixOrZero(int64(t))
	}
	s := strings.TrimSpace(fmt.Sprintf("%v", v))
	if s == "" {
		return time.Time{}, false
	}
	for _, layout := range statTimeLayouts {
		if parsed, err := time.Parse(layout, s); err == nil {
			return parsed, true
		}
	}
	if n, err := strconv.ParseInt(s, 10, 64); err == nil {
		return unixOrZero(n)
	}
	return time.Time{}, false
}

func unixOrZero(n int64) (time.Time, bool) {
	if n <= 0 {
		return time.Time{}, false
	}
	return time.Unix(n, 0).UTC(), true
}

File: web/src/utils/account.ts
// utils/account.ts — the two account deadlines a tracker can quietly miss.
//
// The backend derives login_days_remaining and api_key_expiry_days from a
// stored timestamp plus a clock (internal/stats/account.go) and OMITS them
// whenever it lacks the input. Everything here therefore treats "field absent"
// as "no deadline known" and shows nothing — never a reassuring zero.
import type { TrackerStatsResponse } from '../types';
import { esc } from './format';

/** How near a deadline has to be before the badge appears, in days. */
const LOGIN_WARN_DAYS = 14;
const KEY_WARN_DAYS = 30;

interface AccountWarning {
  kind: 'login' | 'api_key';
  /** Days left; negative means the deadline has passed. */
  days: number;
  /** 'red' once the deadline is past, 'amber' while it can still be met. */
  level: 'amber' | 'red';
  label: string;
  title: string;
}

function num(resp: TrackerStatsResponse | undefined, key: string): number | null {
  const v = resp?.fields?.[key]?.value;
  if (v === null || v === undefined || v === '') return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

function str(resp: TrackerStatsResponse | undefined, key: string): string {
  const v = resp?.fields?.[key]?.value;
  return v === null || v === undefined ? '' : String(v);
}

/** "in 5 days" / "today" / "5 days ago" — the badge's countdown. */
function fmtDays(days: number): string {
  if (days < 0) return `${-days}d overdue`;
  if (days === 0) return 'today';
  return `${days}d`;
}

/** A stored timestamp as a plain calendar date, or '' if it isn't one. */
export function statDate(raw: string): string {
  const s = raw.trim();
  if (!s) return '';
  const d = new Date(s.includes('T') || !s.includes(' ') ? s : s.replace(' ', 'T') + 'Z');
  return isNaN(d.getTime()) ? s : d.toISOString().slice(0, 10);
}

/**
 * The deadline badges for one tracker, nearest first. Empty when nothing is
 * near — which is the normal state, and the reason these are badges rather
 * than a permanent row.
 */
function accountWarnings(resp: TrackerStatsResponse | undefined): AccountWarning[] {
  const out: AccountWarning[] = [];

  const login = num(resp, 'login_days_remaining');
  if (login !== null && login <= LOGIN_WARN_DAYS) {
    const since = num(resp, 'days_since_login');
    const last = statDate(str(resp, 'last_login'));
    // The tracker's own gap in days is not sent to the client, so the tooltip
    // says what IS known — when you last logged in, and how long is left —
    // rather than paraphrasing a policy it would have to guess at.
    const parts = [
      last ? `Last login ${last}` : 'Last login unknown',
      since !== null ? `${since} days ago` : '',
      login < 0
        ? `${-login} days past this tracker's inactivity deadline`
        : `${login} days before this tracker's inactivity deadline`,
    ].filter(Boolean);
    out.push({
      kind: 'login',
      days: login,
      level: login < 0 ? 'red' : 'amber',
      label: login < 0 ? 'Login overdue' : `Login ${fmtDays(login)}`,
      title: parts.join(' · '),
    });
  }

  const key = num(resp, 'api_key_expiry_days');
  if (key !== null && key <= KEY_WARN_DAYS) {
    const on = statDate(str(resp, 'api_key_expires_at'));
    out.push({
      kind: 'api_key',
      days: key,
      level: key < 0 ? 'red' : 'amber',
      label: key < 0 ? 'Key expired' : `Key ${fmtDays(key)}`,
      title: key < 0
        ? `API key expired${on ? ` on ${on}` : ''} — stats stop until you issue a new one`
        : `API key expires${on ? ` on ${on}` : ''} — ${key} days left`,
    });
  }

  return out.sort((a, b) => a.days - b.days);
}

/** Badge markup for a card/row header. '' when nothing is near. */
export function accountWarningBadges(resp: TrackerStatsResponse | undefined): string {
  return accountWarnings(resp)
    .map(w => `<span class="account-warn account-warn--${w.level}" title="${esc(w.title)}">${esc(w.label)}</span>`)
    .join('');
}

/**
 * Whole CALENDAR days from today until `raw`'s day; negative once past. null if
 * the timestamp can't be read.
 *
 * Calendar days, not elapsed hours, because every caller below pairs this with
 * a word — "today", "yesterday", "expired" — and those are claims about the
 * date, not about a duration. Truncating elapsed time got both ends wrong: a
 * login at 23:00 yesterday is -0.4 days, and Math.trunc(-0.4) is -0, so the row
 * read "2026-09-02 · today" on the 3rd. Worse for a key, where -0 < 0 is false
 * in JS, so one that expired six hours ago still announced "expires today".
 *
 * UTC to match statDate, which formats the date sitting next to it.
 */
function calendarDaysUntil(raw: string): number | null {
  const s = raw.trim();
  if (!s) return null;
  const d = new Date(s.includes('T') || !s.includes(' ') ? s : s.replace(' ', 'T') + 'Z');
  if (isNaN(d.getTime())) return null;
  const midnight = (t: Date) => Date.UTC(t.getUTCFullYear(), t.getUTCMonth(), t.getUTCDate());
  return Math.round((midnight(d) - midnight(new Date())) / 86400000);
}

/** "2026-08-30 · 3 days ago" — the Last Login row. */
export function fmtLastLogin(raw: string): string {
  const ago = calendarDaysUntil(raw);
  if (ago === null) return raw;
  return `${statDate(raw)} · ${fmtAgoWords(ago)}`;
}

/** "2027-01-01 · in 121 days" — the API Key Expires row. */
export function fmtKeyExpiry(raw: string): string {
  const left = calendarDaysUntil(raw);
  if (left === null) return raw;
  const rel = left < 0 ? `expired ${-left} day${left === -1 ? '' : 's'} ago`
    : left === 0 ? 'expires today'
    : `in ${left} day${left === 1 ? '' : 's'}`;
  return `${statDate(raw)} · ${rel}`;
}

/** Just the relative half — "today", "yesterday", "4 days ago". Used where the
 *  exact date adds nothing: under a tracker's login policy, the only question
 *  is how the gap so far compares to the gap allowed. */
export function fmtLoginAgo(raw: string): string {
  const ago = calendarDaysUntil(raw);
  return ago === null ? raw : fmtAgoWords(ago);
}

function fmtAgoWords(days: number): string {
  if (days >= 0) return 'today';
  if (days === -1) return 'yesterday';
  return `${-days} days ago`;
}
